sites/website/src/docs/design-systems/creating-a-component-library.md

```diff
@@ -0,0 +1,117 @@
+ => css`/* ... */`;
+const myElementTemplate = (
+context: ElementDefinitionContext,
+definition: FoundationElementDefinition
+) => html`<!-- ... -->`;
+```
+
+Let’s take a closer look at what these two arguments give us.
+
+#### What is the `ElementDefinitionContext`?
+// TODO Rob - I'm not sure how to explain this argument well or what all of it's capabilities are
+
+Importantly, the ElementDefinitionContext can be used to inspect the tag-name for other components using the `ElementDefinitionContext.tagFor()` method. This allows usage of other library components in a template or stylesheet without knowing ahead of time what the tag-name for the element will be:
+
+```ts
+import { html } from "@microsoft/fast-element";
+import { FASTButton } from "@microsoft/fast-components";
+import type { ElementDefinitionContext } from "@microsoft/fast-foundation";
+
+const template = (context: ElementDefinitionContext) => {
+	const buttonTag = context.tagFor(FASTButton);
+
+	return html`<${buttonTag}>Click me!</${buttonTag}>`
+}
+```
+
+#### What is the `FoundationElementDefinition`?
+The `FoundationElementDefinition` is the configuration for which the element was defined, allowing inspection of the aggregated options configuration for the component.
+
+This definition can be extended arbitrarily by components during registration, which is what facilitates default slotted content overrides and arbitrary element configuration:
+
+```ts
+import { html } from "@microsoft/fast-element";
+import { ElementDefinitionContext, FoundationElementDefinition } from "@microsoft/fast-foundation";
+
+interface MyElementDefinition extends FoundationElementDefinition {
+    defaultContent?: string;
+}
+
+const template = (context: ElementDefinitionContext, definition: MyElementDefinition) => {
+    return html`<slot>${definition.defaultContent}</slot>`
+}
+```
+
+During component registration with the Design System, an application author can provide a `defaultContent` field as a configuration, which is used by the template if it exists.
+
+### Compose and Export Registration
+Lastly, create and export the registration function so that application authors can register the component in their Design System:
+
+```ts
+export const myElementRegistration = MyElement.compose({
+  baseName: ‘my-element’,
+  template,
+  styles,
+  defaultContent: "Click me!"
+})
+```
+
+Note that the registration is composed with a "defaultContent" value. This value can be provided during composition or as a conditional in the template; either approach will work.
+
+### Registering Library Components in an Application
+To register the component, an application author will import the registration and register it in their DesignSystem, overriding any properties as necessary:
+
+```ts
+import { myElementRegistration } from "your-package";
+import { DesignSystem } from "@microsoft/fast-foundation";
+
+DesignSystem.getOrCreate().register(
+	myElementRegistration({defaultContent: "Don’t click me!"})
```

Comment:
I think it's OK for App Dev to register a component with a certain prefix, but the rest should be of LD concern, especially all defaults like this

even if this is technically possible and has some rare use-cases, I would not recommend describing this use case without a precaution

Reply:
I'm not sure I agree that this is solely a library's concern - in several cases we've had feedback that folks want to change the default slotted content across the entire app in a single location. That's really the intent behind this. I think it's more common than you might think. For instance, I want to change the default glyph for the tree view item expand/collapse glyph. Without the above API (and corresponding doc) you really can't make that change holistically across the app without either wrapping the component and passing new content via a slot as a new "element", or changing the slotted content everywhere the element shows up.

Reply:
I think this scenario may just need further clarification in the docs.

Reply:
@chrisdholt I didn't say App Devs don't want it :D

not to disagree, but to share my experience: the situation I had many times is when a Feature Dev comes and asks to add a possibility to change smth because his Feature UXer asked to add that or Feature Dev saw it on designs and the library doesn't have the API to achieve it

then you ask a Global UXer (who is usually very strict as his goal is to ensure consistency across the app or multiple apps) and it turns out it was not a good idea to change it on a feature level and the Feature UXer and Global UXer align to prevent the need for deviations from an existing standard

most of the time such alignment results in no change in the library or how library is instantiated in the app, but there are some exceptions, hence I see it as a rare use-case, even though there are parts which are more likely to be accepted for changes on a feature level than others

maybe in this docs a better example is needed to justify the flexibility

Reply:
Totally understand, and I've been on all sides myself :)

I think with regard to our goals, we want to aid in extensibility and enable easily updating or changing for any reason - greater alignment perhaps for some, experimentation with ease for others, etc. I can say with confidence that we have folks across the spectrum you're describing :). I think @EisenbergEffect's call out on clarity will help, and to your point, I think perhaps the example can improve to help articulate a reason someone *may* want to do this. Would that help?

Reply:
I reworked a good portion of the article. Have a read through the latest and see if things are any clearer.